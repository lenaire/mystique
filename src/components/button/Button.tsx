import { forwardRef, useCallback, useEffect, useState, useRef, MouseEvent } from "react";
import styled, { DefaultTheme } from "styled-components";
import { CmpProps } from "$types";
import { Icon, IconType } from "$cmp/icon/Icon";
import { isPromise } from "$utility/promiseFunctions";
import { Direction } from "$utility/enums";

export type ButtonType = "primary" | "secondary";
export type ButtonSize = "sm" | "md" | "lg";

type ClickCallback = (e: MouseEvent<HTMLButtonElement>) => void | Promise<void>;

const getBackgroundColor = (
  theme: DefaultTheme,
  type: ButtonType,
  disabled: boolean | undefined
): string => {
  if (disabled) return theme.palette.grey[400];

  if (type === "secondary") return "transparent";

  // #1ea7fd
  return theme.palette[type].main;
};

export interface ButtonProps extends CmpProps {
  /**
   * Is this the principal call to action on the page?
   */
  type?: ButtonType;
  /**
   * How large should the button be?
   */
  size?: ButtonSize;
  /**
   * Button contents
   */
  label: string;
  submittingLabel?: string;
  disabled?: boolean;
  isSubmit?: boolean;
  direction?: Direction;
  /**
   * Optional click handler
   */
  onClick?: ClickCallback;
}

const Container = styled.button<{
  $size: ButtonSize;
  $type: ButtonType;
  $disabled: boolean;
}>`
  height: ${(props): string => `${props.theme.palette.button[props.$size].height}px`};
  padding: ${(props): string => `${props.theme.palette.button[props.$size].padding}`};
  font-size: ${(props): string => `${props.theme.palette.button[props.$size].fontSize}pt`};
  color: ${(props): string => `${props.$type === "primary" ? "#fff" : " #333"}`};
  background-color: ${(props): string =>
    getBackgroundColor(props.theme, props.$type, props.$disabled)};
  box-shadow: ${(props): string =>
    `${props.$type === "primary" ? "" : "rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset"}`};
  font-family: "Nunito Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
  font-weight: 700;
  font-size: ${(props): string => `${props.theme.palette.button[props.$size].fontSize}pt`};
  border: 0;
  border-radius: 3em;
  cursor: pointer;
  display: inline-block;
  line-height: 1;
  cursor: ${(props): string | undefined => (props.$disabled ? "default" : undefined)};
`;

/**
 * Primary UI component for user interaction
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      type = "primary",
      size = "md",
      label,
      onClick,
      submittingLabel,
      isSubmit,
      disabled,
      className,
      ...rest
    },
    ref
  ) => {
    const [isSubmitting, setIsSubmitting] = useState(false);

    const isMountedRef = useRef(true);
    useEffect(() => {
      isMountedRef.current = true;
      return (): void => {
        isMountedRef.current = false;
      };
    }, []);

    const asyncOnClick = useCallback(
      (event: MouseEvent<HTMLButtonElement>) => {
        if (!isSubmit) {
          event.preventDefault();
        }

        if (onClick) {
          const result = onClick(event);

          if (isPromise(result)) {
            setIsSubmitting(true);
            result
              .then(() => {
                if (isMountedRef.current) {
                  setIsSubmitting(false);
                }
              })
              .catch((e) => {
                if (isMountedRef.current) {
                  setIsSubmitting(false);
                }

                throw e;
              });
          }
        }
      },
      [onClick, isSubmit]
    );

    const handleClick = ((): ClickCallback | undefined => {
      if (disabled) return undefined;

      return asyncOnClick;
    })();

    return (
      <Container
        type={isSubmit ? "submit" : "button"}
        className={className}
        ref={ref}
        $type={type}
        $size={size}
        $disabled={disabled || isSubmitting}
        onClick={handleClick}
        {...rest}
      >
        {/* TODO: Typography component */}
        {isSubmitting ? submittingLabel : label}
      </Container>
    );
  }
);
