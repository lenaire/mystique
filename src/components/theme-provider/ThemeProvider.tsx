import { FC } from "react";
import {
  DefaultTheme, // This is overridden in types/theme.d.ts
  ThemeProvider as StyledComponentsThemeProvider,
  ThemeProviderProps as StyledComponentsThemeProviderProps,
} from "styled-components";

import { defaultTheme } from "$cmp/theme-provider/theme";
import { ResetGlobalCSS } from "$cmp/theme-provider/ResetGlobalCSS";

export type ThemeProviderProps = Partial<StyledComponentsThemeProviderProps<DefaultTheme>>;

export const ThemeProvider: FC<ThemeProviderProps> = ({ children, theme = defaultTheme }) => {
  return (
    <StyledComponentsThemeProvider theme={theme}>
      <ResetGlobalCSS />
      {children}
    </StyledComponentsThemeProvider>
  );
};
