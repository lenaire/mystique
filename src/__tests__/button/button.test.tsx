import renderer from "react-test-renderer";
import { ThemeProvider } from "$cmp/theme-provider/ThemeProvider";
import { Button, ButtonProps } from "$cmp/button/Button";

describe("<Button />", () => {
  it("renders correctly", () => {
    const props: ButtonProps = {
      label: "Continue",
    };

    const tree = renderer
      .create(
        <ThemeProvider>
          <Button {...props} />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
