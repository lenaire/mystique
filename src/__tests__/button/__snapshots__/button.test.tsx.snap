// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<button
  className="sc-bcXHqe iJirDr"
  onClick={[Function]}
  type="button"
>
  Continue
</button>
`;
